Implement crypto_secretstream as IncrementalXChaCha20Poly1305

Jack-Edwards opened the pull request:
I am not very comfortable with references and pointers in C#, but I did my best to copy your patterns.

I'm going to leave some comments in the PR in some areas I'm unsure of.

Jack-Edwards committed:
Implement crypto_secretstream as IncrementalXChaCha20Poly1305

Jack-Edwards commented on src/Geralt.Tests/XChaCha20Poly1305Tests.cs:
```diff
@@ -183,4 +183,133 @@ public void Decrypt_InvalidKey()
     }
+
+    [TestMethod]
+    public void IncrementalEncryptDecrypt_ValidInputs()
```
I had to combine encryption and decryption in the same unit tests.  As the libsodium documentation states, the same sequence encrypted twice will produce different ciphertexts.

Jack-Edwards commented on src/Geralt/Interop/Interop.XChaCha20Poly1305.cs:
```diff
@@ -8,10 +8,42 @@ internal static partial class Libsodium
+        [DllImport(DllName, CallingConvention = Convention)]
+        internal static extern unsafe void crypto_secretstream_xchacha20poly1305_rekey(ref crypto_secretstream_xchacha20poly1305_state state);
+
+        [StructLayout(LayoutKind.Explicit, Size = 52)]
```
I pulled `52` from the struct in the libsodium repo.  Things are working fine, but I'd appreciate a sanity check.

samuel-lucas6 replied:
Seems right to me too.

Jack-Edwards commented on src/Geralt/Interop/Interop.XChaCha20Poly1305.cs:
```diff
@@ -8,10 +8,42 @@ internal static partial class Libsodium
+        internal const int crypto_secretstream_xchacha20poly1305_KEYBYTES = 32;
+
+        // Not sure how or whether to handle this
+        // internal const long crypto_secretstream_xchacha20poly1305_MESSAGEBYTES_MAX = big number;
```
I don't know that ~256 GB is worth validating for.  But I also don't know how to even interpret the arithmetic in the libsodium repo.

samuel-lucas6 replied:
Those commented lines can be removed. libsodium does a check internally, and you can't pass in a message that large, so I'm not checking for ChaCha20-Poly1305 either.

Jack-Edwards committed:
Additional unit tests for IncrementalXChaCha20Poly1305

samuel-lucas6 commented:
This was a pleasant surprise! Thanks Jack! I will try to take a look today.

Jack-Edwards committed:
IncrementalXChaCha20Poly1305 key should always be provided

Jack-Edwards commented:
Thanks.  I'm looking into alternatives to libsodium-core and I think Geralt is the closest to what I'm looking for.  crypto_secretstream support may be the only thing missing for my use-case.

samuel-lucas6 commented:
> I'm looking into alternatives to libsodium-core and I think Geralt is the closest to what I'm looking for.

That's good to hear. Please feel free to share any feedback on the API and documentation. I need to rewrite all the tests to try and reduce repetition at some point, and I'd like to change the BLAKE2b stream implementation so it uses spans.

> crypto_secretstream support may be the only thing missing for my use-case.

Yes, I think it makes sense now there's IncrementalBLAKE2b and Poly1305.

samuel-lucas6 commented on src/Geralt/Crypto/IncrementalXChaCha20Poly1305.cs:
```diff
@@ -0,0 +1,118 @@
+    public const int KeySize = crypto_secretstream_xchacha20poly1305_KEYBYTES;
+
+    /// <summary>
+    /// No information is added to the nature of the message.
```
'about' instead of 'to'.

samuel-lucas6 commented on src/Geralt/Crypto/IncrementalXChaCha20Poly1305.cs:
```diff
@@ -0,0 +1,118 @@
+    /// <summary>
+    /// No information is added to the nature of the message.
+    /// </summary>
+    public const byte TagMessage = crypto_secretstream_xchacha20poly1305_TAG_MESSAGE;
```
I'm concerned calling these tags may be confusing because they're different to the Poly1305 tag. What do you think of `FlagMessage`, `FlagFinal`, etc?

samuel-lucas6 replied:
`ref byte tag` would also need to be changed.

Jack-Edwards replied:
I prefer "tag" over "flag", but I'm not opposed to an alternative.

Jack-Edwards replied:
I'm toying around with switching the bytes over to an enum.  Calling the enum `StreamFlag`.

samuel-lucas6 commented on src/Geralt/Crypto/IncrementalXChaCha20Poly1305.cs:
```diff
@@ -0,0 +1,118 @@
+    /// <summary>
+    /// Size of the header, which gets created when initializing for encryption and must be provided back when initializing for decryption.
+    /// </summary>
+    public const int HeaderBytes = crypto_secretstream_xchacha20poly1305_HEADERBYTES;
```
`HeaderSize` to be consistent. Everything operates on bytes.

samuel-lucas6 commented on src/Geralt/Crypto/IncrementalXChaCha20Poly1305.cs:
```diff
@@ -0,0 +1,118 @@
+    /// </summary>
+    public const byte TagReKey = crypto_secretstream_xchacha20poly1305_TAG_REKEY;
+
+    public IncrementalXChaCha20Poly1305(bool forDecryption, Span<byte> header, ReadOnlySpan<byte> key)
```
Just `decryption` instead of `forDecryption`?

samuel-lucas6 replied:
`private bool _forDecryption;` would also need to be changed.

samuel-lucas6 commented on src/Geralt/Crypto/IncrementalXChaCha20Poly1305.cs:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    public unsafe void Push(Span<byte> ciphertextChunk, ReadOnlySpan<byte> plaintextChunk, ReadOnlySpan<byte> associatedData, byte tag = TagMessage)
```
Can we merge the no associated data and associated data methods (`Push` and `Pull`) to save some code? I understand why you've done it though; `ReadOnlySpan<byte> associatedData = default` might have to go at the end in that case.

Jack-Edwards replied:
Yes, done.

samuel-lucas6 commented on src/Geralt/Crypto/IncrementalXChaCha20Poly1305.cs:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    public unsafe void ReKey()
```
`Rekey` looks nicer to me.

samuel-lucas6 commented on src/Geralt/Crypto/IncrementalXChaCha20Poly1305.cs:
```diff
@@ -0,0 +1,118 @@
+
+    public unsafe void Pull(Span<byte> plaintextChunk, ReadOnlySpan<byte> ciphertextChunk, ReadOnlySpan<byte> associatedData, ref byte tag)
+    {
+        if (!_forDecryption) { throw new InvalidOperationException("Cannot pull from an encryption stream."); }
```
It might be worth moving this check to the `Validation` class like `EqualToSize`. I can do that later though.

samuel-lucas6 commented on src/Geralt/Crypto/IncrementalXChaCha20Poly1305.cs:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    public unsafe void Pull(Span<byte> plaintextChunk, ReadOnlySpan<byte> ciphertextChunk, ref byte tag)
```
Perhaps `Pull` should `return tag;` instead of `ref byte tag`?

samuel-lucas6 commented on src/Geralt/Crypto/IncrementalXChaCha20Poly1305.cs:
```diff
@@ -0,0 +1,118 @@
+            int ret = _forDecryption
+                ? crypto_secretstream_xchacha20poly1305_init_pull(ref _state, h, k)
+                : crypto_secretstream_xchacha20poly1305_init_push(ref _state, h, k);
+            if (ret != 0) { throw new CryptographicException("Error initializing stream encryption/decryption."); }
```
I've spelt it 'initialising' in other errors because I'm British.

Jack-Edwards replied:
That's a shame :)

samuel-lucas6 commented on src/Geralt.Tests/XChaCha20Poly1305Tests.cs:
```diff
@@ -183,4 +183,208 @@ public void Decrypt_InvalidKey()
+        Span<byte> header = stackalloc byte[IncrementalXChaCha20Poly1305.HeaderBytes];
+        Span<byte> plaintext = stackalloc byte[Plaintext.Length];
+        Span<byte> ciphertext = stackalloc byte[Plaintext.Length + IncrementalXChaCha20Poly1305.TagSize];
+        byte outTag = 0xFF;
```
`outTag` will need renaming everywhere in this file if the variables get changed in the other file.

samuel-lucas6 commented on src/Geralt.Tests/XChaCha20Poly1305Tests.cs:
```diff
@@ -183,4 +183,208 @@ public void Decrypt_InvalidKey()
+        Span<byte> plaintext = stackalloc byte[Plaintext.Length];
+        Span<byte> ciphertext = stackalloc byte[Plaintext.Length + IncrementalXChaCha20Poly1305.TagSize];
+        byte outTag = 0xFF;
+        using var encryptor = new IncrementalXChaCha20Poly1305(false, header, Key);
```
Please can you add `decryption: false` and `decryption: true` everywhere.

Jack-Edwards replied:
I missed this.  Adding it now.

samuel-lucas6 commented on src/Geralt.Tests/XChaCha20Poly1305Tests.cs:
```diff
@@ -183,4 +183,208 @@ public void Decrypt_InvalidKey()
+        using var encryptor = new IncrementalXChaCha20Poly1305(false, header, Key);
+        encryptor.Push(ciphertext, Plaintext, IncrementalXChaCha20Poly1305.TagFinal);
+        using var decryptor = new IncrementalXChaCha20Poly1305(true, header, Key);
+        decryptor.Pull(plaintext, ciphertext, tag: ref outTag);
```
I'd remove `tag:` because the variable naming is clear.

samuel-lucas6 commented on src/Geralt/Crypto/IncrementalXChaCha20Poly1305.cs:
```diff
@@ -0,0 +1,118 @@
+    /// <summary>
+    /// "Forget" the current secret key and derive a new one.
+    /// </summary>
+    public const byte TagReKey = crypto_secretstream_xchacha20poly1305_TAG_REKEY;
```
I'd rename this to `TagRekey`. Furthermore, these const bytes might be better of as an enum to prevent the user passing other bytes in.

Jack-Edwards committed:
Remove unnecessary comments from interop code

samuel-lucas6 commented on src/Geralt.Tests/XChaCha20Poly1305Tests.cs:
```diff
@@ -183,4 +183,208 @@ public void Decrypt_InvalidKey()
+    {
+        Span<byte> header = stackalloc byte[IncrementalXChaCha20Poly1305.HeaderBytes];
+        int plaintext1Length = 50;
+        int plaintext2Length = 50;
```
I'd put these on the same line everywhere given they have the same value.

samuel-lucas6 commented on src/Geralt.Tests/XChaCha20Poly1305Tests.cs:
```diff
@@ -183,4 +183,208 @@ public void Decrypt_InvalidKey()
+        using var decryptor = new IncrementalXChaCha20Poly1305(true, header, Key);
+        decryptor.Pull(plaintext[0..plaintext1Length], ciphertext1, tag: ref outTag);
+        Assert.AreEqual(IncrementalXChaCha20Poly1305.TagMessage, outTag);
+        // decryptor.ReKey(); // Where the rekey should have occurred
```
This comment can be removed due to the method name.

samuel-lucas6 commented on src/Geralt.Tests/XChaCha20Poly1305Tests.cs:
```diff
@@ -183,4 +183,208 @@ public void Decrypt_InvalidKey()
+        Span<byte> header = stackalloc byte[IncrementalXChaCha20Poly1305.HeaderBytes];
+        Span<byte> plaintext = stackalloc byte[Plaintext.Length];
+        Span<byte> ciphertext = stackalloc byte[Plaintext.Length + IncrementalXChaCha20Poly1305.TagSize];
+        byte[] associatedData = new byte[] { 0x01, 0x02, 0x03, 0x04 };
```
This can be `Span<byte>` with stackalloc.

samuel-lucas6 commented on src/Geralt.Tests/XChaCha20Poly1305Tests.cs:
```diff
@@ -183,4 +183,208 @@ public void Decrypt_InvalidKey()
+    }
+
+    [TestMethod]
+    public void IncrementalEncryptDecrypt_ValidInputs_WithReKey()
```
We can probably get away with merging `Rekey()` into the currently non-rekeying test to save some lines.

Jack-Edwards replied:
I prefer keeping this test.  If this test fails we know the issue has something to do with rekey and not something else.

samuel-lucas6 reviewed (changes requested):
I can make these changes if you'd prefer, although I'd appreciate your feedback on some of them.

Jack-Edwards committed:
Changes based on feedback

Jack-Edwards committed:
Combine duplicate Push and Pull methods

Jack-Edwards commented:
@samuel-lucas6 I implemented most of your feedback and resolved some compiler messages I hadn't seen before.  I'll leave it up to you to resolve any open conversations.

Edit: Okay *now* I've implemented most of your feedback.  Some of your comments were only showing up under Files Changed.

Jack-Edwards committed:
Used named "decryption" parameter in IncrementalXChaCha20Poly1305 unit tests

Jack-Edwards committed:
Update error message to be more British

samuel-lucas6 committed:
Test renaming

samuel-lucas6 commented:
Appreciate it. Last thing is could you merge the no associated data and associated data methods for `Pull` like you did for `Push`? Not a fan of the GitHub editor. Then I'll merge.

Jack-Edwards committed:
Remove overload for IncrementalXChaChaPoly1305 'Pull' method

Jack-Edwards commented:
The overload for Pull is gone.
